[marcellaantania] ip

marcellaantania opened the pull request:
## Thea
> “Improved productivity means less human sweat, not more.” –Henry Ford. ([Source](https://emailanalytics.com/101-productivity-quotes-from-historys-most-brilliant-minds/))

_The name Thea comes from the Greek word "[Alethea](https://www.dictionary.com/browse/alethea)" which means "truth"._

Want to increase your productivity by listing out your tasks in one place?
Thea is here to help you! :smiling_face_with_three_hearts:


Why should you use Thea, you ask?
- Text-based
- Easy to use
- Simple and effective
- Cute personality to become your ~~only~~ friend

What can Thea do? 
- [x] Help you remember your tasks and deadlines
- [x] Help you remember your events
- [x] Remember your tasks until the next time you call for Thea again
- [x] Motivate you to finish your task 💪 

Have you been persuaded? 😎 Follow the steps below!
1. Download the files from [here](https://github.com/marcellaantania/ip).
2. Double-click it.
3. Add your tasks.
4. Let Thea manage your tasks for you! 😉 

The best part is? It is **FREE** to use!

What are you waiting for!! Never miss a single task anymore from now on with the help of Thea! 🥇 

---

If you are a developer, feel free to check out the main logic behind Thea by checking out the `Thea` class in the repository. Here is the constructor of Thea for a sneak peek:
```ruby
public class Thea {
    private Storage storage;
    private TaskList tasks;
    private Ui ui;

    public Thea(String fileName) {
        this.ui = new Ui();
        this.storage = new Storage(fileName);
        try {
            this.tasks = new TaskList(storage.retrieveTasks());
        } catch (FileCorruptedException e) {
            ui.showError(e.getMessage());
            tasks = new TaskList();
        }
    }
}
```

Commit:
Add Gradle support

Eclipse-Dominator committed:
Bump gradle and lib version

marcellaantania committed:
Name change from Duke to Thea, changed greeting and exiting message

marcellaantania committed:
Merge remote-tracking branch 'origin/add-gradle-support'

marcellaantania committed:
echo function

marcellaantania committed:
add, list function

marcellaantania committed:
added Task class

marcellaantania committed:
integrate Thea with Task class, add mark/unmark function

marcellaantania committed:
Made three different subclasses of Task (ToDo, Deadline, Event)

marcellaantania committed:
Fix typo

marcellaantania committed:
change access modifier

marcellaantania committed:
update Thea to support 3 types of tasks: ToDo, Deadline and Event

marcellaantania committed:
remove extra nextline

marcellaantania committed:
remove emoji that cannot be run

marcellaantania committed:
Automated Testing

marcellaantania committed:
Thea-specific exceptions

marcellaantania committed:
Update code to throw three types of thea-specific exceptions

marcellaantania committed:
Add delete function

marcellaantania committed:
Read file functionality

marcellaantania committed:
Add an exception for unexpected file format

marcellaantania committed:
Update Thea to retrieve data as tasks

marcellaantania committed:
Write tasks to hard disk, not fully functional yet

marcellaantania committed:
Update Thea and subclasses of Task to support task saving

marcellaantania committed:
Merge branch 'branch-Level-7'

marcellaantania committed:
Save deadline duedates and event timings as LocalDateTime

marcellaantania committed:
Merge branch 'branch-Level-8'

bkjwjason commented on src/main/java/Thea.java:
```diff
@@ -0,0 +1,181 @@
+            ArrayList<String> commandWordsArray = new ArrayList<>();
+            Collections.addAll(commandWordsArray, commandWords);
+            String command = commandWords[0];
+            if (command.equals("bye")) {
+                exit();
+                break;
+            } else if (command.equals("list")) {
+                printList(tasks);
+            } else if (command.equals("mark")) {
+                int index = Integer.parseInt(commandWords[1]) - 1;
+                if (index > tasks.size() - 1) {
+                    throw new IndexOutOfBoundsException("There is currently no task " + (index + 1));
+                } else {
+                    tasks.get(index).markAsDone();
+                    System.out.printf("Great job! ˊᗜˋ I've marked this task as done:\n  %s\n", tasks.get(index));
+                }
+            } else if (command.equals("unmark")) {
+                int index = Integer.parseInt(commandWords[1]) - 1;
+                if (index > tasks.size() - 1) {
+                    throw new IndexOutOfBoundsException("There is currently no task " + (index + 1));
+                } else {
+                    tasks.get(index).unmarkAsDone();
+                    System.out.printf("Okay, I've marked this task as not done yet:\n  %s\n", tasks.get(index));
+                }
+            } else if (command.equals("delete")){
```
An alternative less tedious way might be using switch/case statements?

marcellaantania replied:
I have updated the code! Thank you for your suggestion.

bkjwjason commented on src/main/java/Thea.java:
```diff
@@ -0,0 +1,181 @@
+                    throw new IndexOutOfBoundsException("There is currently no task " + (index + 1));
+                } else {
+                    System.out.println("I have removed the following task to your list:\n  "
+                            + tasks.get(index).toString() + "\nNow you have " + (tasks.size() - 1)
+                            + ((tasks.size() - 1) == 1 ? " task" : " tasks")
+                            + " in the list.");
```
Great job on adhering to code style by bringing operator to next line!

bkjwjason commented on src/main/java/Thea.java:
```diff
@@ -0,0 +1,181 @@
+        String line;
+        ArrayList<Task> retrievedTasks = new ArrayList<>();
+        Task task;
+        try (BufferedReader bufferReader = Files.newBufferedReader(path)) {
+            line = bufferReader.readLine();
+            while (line != null) {
+                String[] splitLine = line.split(" [|] ");
+                if (splitLine[0].equals("T")) {
+                    task = new ToDo(splitLine[2]);
+                } else if (splitLine[0].equals("D")) {
+                    task = new Deadline(splitLine[2], splitLine[3]);
+                } else if (splitLine[0].equals("E")) {
+                    task = new Event(splitLine[2],splitLine[3], splitLine[4]);
+                } else {
+                    throw new FileCorruptedException("Unexpected File Format Found. File might be corrupted.");
+                }
+                if (splitLine[1].equals("1")) {
+                    task.markAsDone();
+                }
+                retrievedTasks.add(task);
+                line = bufferReader.readLine();
+            }
+        } catch (IOException e) {
+            throw new RuntimeException(e);
```
Do consider abstracting out the parsing into a separate function to avoid "arrowhead" style code!

bkjwjason commented on src/main/java/Thea.java:
```diff
@@ -0,0 +1,181 @@
+
+public class Thea {
+
+    public static void main(String[] args) throws Exception {
```
Maybe you can consider abstracting out parsing into a separate class to keep your main function shorter?

marcellaantania replied:
Thank you for your review! I have done so and pushed the new code into the PR. Feel free to check the newest code update if you are interested :)

bkjwjason reviewed (commented):
Overall, the code had minor styling issues. What would be great is if you abstract out certain classes like Parser and UI such that each method wouldn't be too long. Good job!

marcellaantania committed:
Add a new exception to handle wrong date and time format from the user input

marcellaantania committed:
Extract classes from the main Thea class

marcellaantania committed:
Add parser and command class to understand user input and return them as commands

marcellaantania committed:
Modify Thea class to use the extracted classes

marcellaantania committed:
Merge branch 'branch-A-MoreOOP'

marcellaantania committed:
Package everything to the package "thea"

marcellaantania committed:
Merge branch 'A-Packages'

marcellaantania committed:
Update gradle to work with Thea

marcellaantania committed:
Merge branch 'A-Gradle'

marcellaantania committed:
Add test class with dummy tests

marcellaantania committed:
Add unit tests for ToDo, Deadline and Event classes

marcellaantania committed:
Remove the dummy TheaTest class

marcellaantania committed:
Merge branch 'A-JUnit'

marcellaantania committed:
Add jar filename

marcellaantania committed:
Add JavaDoc to all classes.

marcellaantania committed:
Edit to follow java coding standard

marcellaantania committed:
Add keyword find functionality

marcellaantania committed:
Merge branch 'branch-A-JavaDoc'

marcellaantania committed:
Merge branch 'branch-A-CodingStandard'

# Conflicts:
#	src/main/java/thea/DeleteCommand.java

marcellaantania committed:
Merge branch 'branch-Level-9'

# Conflicts:
#	src/main/java/thea/Parser.java

marcellaantania committed:
Add JavaDoc for the new Find functionality

marcellaantania committed:
Change method name to adhere to standard

marcellaantania committed:
Add checkstyle

marcellaantania committed:
Merge branch 'A-Checkstyle'

marcellaantania committed:
Add JavaFX (unfinished)

marcellaantania committed:
Add basic working GUI

marcellaantania committed:
Crop display pictures to circle, add padding to the text

marcellaantania committed:
Change GUI style

marcellaantania committed:
Merge branch 'branch-Level-10'

marcellaantania committed:
Changes to adhere to checkstyle

marcellaantania committed:
Tidy up some code

marcellaantania committed:
Assert that file is created properly

marcellaantania committed:
Storage Class: extract methods

The retrieveTask and saveTaskList methods were too long and had multiple levels of abstraction (Not adhering to SLAP).

Let's,
* Extract code relating to parsing of retrieved data in the retrieveTasks method a new method called parseTask.
* Extract code to create new directory and/or path from the saveTaskList method to a new method called createFileIfNotExist.

Extracting the intermediate steps with different levels of abstraction improves readability, shortens the methods and improves the code in adhering to SLAP.

marcellaantania committed:
MainWindow Class: extract method getResponse and exitProgramIfRequested

The handleUserInput method had multiple levels of abstraction which does not adhere to SLAP. It also hides the happy path of the code.

Let's,
* Extract code relating to parsing of the input to method getResponse.
* Extract code to exit program after a delay if requested by the user.

Extracting the intermediate steps with different levels of abstraction improves readability, shortens the methods and improves the code in adhering to SLAP.

marcellaantania committed:
Parser class: extract method createDeadline, createEvent and checkDateValidity, tidy up redundancies.

The parse methods were way too long, hard to read and had multiple levels of abstraction (Not adhering to SLAP).

Let's,
* Extract code relating to reformatting and creating the Deadline class before it is added as a parameter to the AddCommand.
* Extract code relating to reformatting and creating the Event class before it is added as a parameter to the AddCommand.
* Extract code that checks if the date inputted by the user is correctly formatted. The similar codes were adjusted to work with this one method.
* Combine commandWords with commandWordsArray and used ArrayList directly instead (making use of List.of).

Extracting the intermediate steps with different levels of abstraction improves readability, shortens the methods and improves the code in adhering to SLAP.
Combining separate date format checks into one minimizes code duplication.
As commandWords and commandWordsArray contain the same data, just directly using ArrayList is better as we need the size function.

marcellaantania committed:
Merge pull request #3 from marcellaantania/branch-A-Assertion

Assert that file is created properly

marcellaantania committed:
Merge branch 'master' into branch-A-CodeQuality

marcellaantania committed:
Merge pull request #2 from marcellaantania/branch-A-CodeQuality

Branch a code quality

marcellaantania committed:
Ignore data directory for version control

marcellaantania committed:
Ignore thea.txt

marcellaantania committed:
Add load function

The load function enables user to specify which file they want as the data source.

marcellaantania committed:
Merge branch 'branch-BCD-Extension'

marcellaantania committed:
Update README.md

marcellaantania committed:
Fix typo

marcellaantania committed:
Add application screenshot (Ui.png)

marcellaantania committed:
Merge branch 'master' of https://github.com/marcellaantania/ip

marcellaantania committed:
Change the main entry of the application in the build.gradle file
